### PR TITLE
os/drivers/serial: Avoid Board Sleep in UART Tx

### PR DESCRIPTION
**Pull request by gSahitya-samsung**
This commit make use of pm_suspend and pm_resume APIs to avoid board sleep during uart_write operation. It register PM_UART_DOMAIN during uart_open to access PM APIs. The prevention of board sleep is required to avoid data loss during uart_write operation.

### CODE REVIEW

**Review by ewoodev — APPROVED** [review]

File: os/drivers/serial/serial.c [hunk 1]
```diff
@@ -1214,6 +1235,18 @@ static int uart_open(FAR struct file *filep)
...
 
+#ifdef CONFIG_PM
+	/* Register PM_UART_DOMAIN to access PM APIs during UART operations. */
+	if (pm_uart_domain_id == -1) {
+		ret = pm_domain_register(PM_UART_DOMAIN);
+		if (ret < 0) {
+			return ret;
+		} else {
+			pm_uart_domain_id = ret;
+		}
```
Comment (ewoodev) [inline os/drivers/serial/serial.c:1246]:
```suggestion
pm_uart_domain_id = ret;
```

it has return, we don't need to eles

**Comment by sunghan-chang** [comment]
@gSahitya-samsung If we registet UART tx IRQ as a wakeup source, what happen?
@ziyik Is it possible? (Sorry, I don't know Dane's user id)

**Comment by sunghan-chang** [comment]
@gSahitya-samsung The title is too long. How about `make pm suspend in uart tx`?

**Comment by ziyik** [comment]
> @gSahitya-samsung If we registet UART tx IRQ as a wakeup source, what happen? @ziyik Is it possible? (Sorry, I don't know Dane's user id)

It does no make sense as register UART tx IRQ as a wakeup source, because when there is something to send the device should be in as wake condition since the data is send from the application or kernel.

**Comment by ziyik** [comment]
@sunghan-chang, I have fixed an issue related to my previous PR, which is more related to the driver side.
Please refer to https://github.com/Samsung/TizenRT/pull/6407

**Comment by sunghan-chang** [comment]
> @gSahitya-samsung The title is too long. How about `make pm suspend in uart tx`?

@gSahitya-samsung After modifying the title, let's merge. Could you check my comment?

**Comment by sunghan-chang** [comment]
> > @gSahitya-samsung If we registet UART tx IRQ as a wakeup source, what happen? @ziyik Is it possible? (Sorry, I don't know Dane's user id)
> 
> It does no make sense as register UART tx IRQ as a wakeup source, because when there is something to send the device should be in as wake condition since the data is send from the application or kernel.

@ziyik Thank you for opinion.

**Comment by gSahitya-samsung** [comment]
> > @gSahitya-samsung The title is too long. How about `make pm suspend in uart tx`?
> 
> @gSahitya-samsung After modifying the title, let's merge. Could you check my comment?

Done

**Review by sunghan-chang — APPROVED** [review]

**Review by ewoodev — APPROVED** [review]

**Comment by sunghan-chang** [comment]
@gSahitya-samsung hmmmm. much minor but.. you also use pm_resume. And using pm_suspend is too surface(?) title? It does not represent what you modify. You want to keep wakeup state at tx operation instead of entering pm sleep.

**Comment by sunghan-chang** [comment]
@gSahitya-samsung Let me suggest `avoid board sleep in UART Tx`. "use pm api" is no need.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**